### ISSUE TITLE
Consider libc

### ISSUE THREAD

**Issue by krasin** [llvm-dcpu16/llvm-dcpu16#42]
One suggestion is http://www.nongnu.org/avr-libc/ and http://www.nongnu.org/avr-libc/user-manual/malloc.html

It's BSD-licensed and at least malloc stuff can be used for DCPU16.

**Comment by ghost** [comment]
I will try to port newlib (http://sourceware.org/newlib/) as soon as possible. It's designed to be portable and it defines which syscalls needs to be defined / remaded in it's documentations. Oh and it's essentialy BSD-license.

**Comment by krasin** [comment]
There's a chance that newlib is a bit bloated, but I see that some people use it on 16-bit acrhitectures, so let's see.

**Comment by theevocater** [comment]
Getting an implementation of malloc depends somewhat on how Notch is going to lay out memory.  I think we can assume that there won't be virtual memory, and he has made mention of there being a set of memory mapped video memory.  

That said, malloc is a big part of getting actual c code going.

**Comment by ghost** [comment]
From screenshots notch has released the memory mapped IO will start at 0x8000. There has been talk about that at the 0x10c forum:

http://0x10cforum.com/forum/m/4932880/viewthread/2711023-preliminary-memory-map-for-dcpu16

**Comment by theevocater** [comment]
thanks! I didn't even know about that.  I'll peruse the forums to catch up on the latest info.

**Comment by krasin** [comment]
There was a comment suggesting http://www.fefe.de/dietlibc/ (I'm not sure where this comment is now, though).

I have no opinion, but I would like to keep it as an option. [The wiki page about dietlibc](http://en.wikipedia.org/wiki/Dietlibc)

**Comment by ghost** [comment]
Dietlibc seems to be GPL only (bad for static linking) and might need a kernel that support it. But haven't looked deeper into it.

**Comment by ghost** [comment]
Okay guys, I think we have everything that we need to implement a very basic libc!
We have a basic support for pointer arithmetic that is based word addressing, we support 16bit chars/bytes and we implemented structs. Since the nos linker also supports the _end symbol, we can now address the address space behind the code, which should make a malloc implementation possible.

**Comment by krasin** [comment]
[Basic malloc implementation](https://github.com/Niriel/dcpu16/blob/master/malloc.dasm16). Unfortunately, its license is unclear.

**Comment by ghost** [comment]
I think we should implement it in C and should orientate on AVR-libc on doing so. Simply because the AVR code has a rather simple, but still smart way to log freed memory space. BUT we should pay special attention to the memory map that is already known: Heap data only between code end and 0x8000 and between 0x9010 and stack end.

ASM is very hard to maintain and I'm sure that LLVM will produce good code for us.

**Comment by krasin** [comment]
yes, you have the point

btw, have you seen [this Notch's tweet](https://twitter.com/#!/notch/status/194836858667155456)?

```
To annoy everyone, I'm switching the names of a and b values in the dcpu-16 specification. 
The new emu runs my test program again!
```

I assume that yes, because you have replied to Notch to his previous tweet.

**Comment by ghost** [comment]
Oh, I didn't saw till now. That's ... work to do for us. But at least we have the default AT&T order GAS uses.

**Comment by krasin** [comment]
In fact, his tweet might mean two things: he changes a and b on the ASM level, or he has changed the binary encoding.
Let's wait for the updated specs.

**Comment by Blei** [comment]
So, has anyone started to work on this already? I reckon what we need is
- malloc
- some helper functions to do interrupt related stuff
- libraries to communicate with the released hardware

**Comment by Blei** [comment]
For everyone interested, here are some utility libs: https://github.com/Blei/DCPU16-Utility-Libs

Depends on #155 to produce reasonable asm.

**Comment by ghost** [comment]
_mhh_ Don't you want to rename it to dcpu16-libc and push all those libs into a "hardware" folder? (and remove those _lib postfix) Then we just need a nice header and we can start making a nice libc alternative. (Which we can also publish in this organisation).

**Comment by Blei** [comment]
These files were just the outcome of me playing around with `asm` a bit :P I didn't think about organisation and design yet. If you want to restructure/rename stuff, I can give you commit rights, or you can open a new repo ("fresh start") and copy the code over if you want.

**Comment by ghost** [comment]
Okay, thank you. I started porting the malloc() from the libc-avr library. But right now, I don't have any "environment" to test my code against. I really need a good assembler/linker and a compatible emulator...

**Comment by aubreyrjones** [comment]
@arbaal assembler/linker in a few days.

But, I don't think I'm going to be able to also maintain an emulator. There has to be somebody who has built an emulator on the latest notch core.

**Comment by ghost** [comment]
You got any link to that emulator?

**Comment by aubreyrjones** [comment]
@arbaal No, I didn't mean that I know of such an emulator. I just figure that there _must_ be somebody who has done so. Or who's hacking on one and will release it fairly soon.
